### ISSUE TITLE
[Boost] static link is broken

### ISSUE THREAD

**Issue by cenit** [Microsoft/vcpkg#1338]
As from the subject, static linking of boost libraries is currently broken.
This commit https://github.com/Microsoft/vcpkg/commit/7811ef602ecd963ae333e3427c82629d1c8e9f34 is the source of the problems: why renaming static libraries, if those are the names that CMake is then really looking for?
As an example, now we rename 
```
x86-windows-static/libboost_thread-vc140-mt-s-1_64.lib
```
into
```
x86-windows-static/boost_thread-vc140-mt-1_64.lib.
```
But this is wrong: boost_thread-vc140-mt-1_64.lib is known to CMake to be the filename of an import library (allowing dynamic linking with boost_thread-vc140-mt-1_64.dll), while *libboost_thread-vc140-mt-s-1_64.lib* should be the name of the static library (also with the -s, meaning it's statically linked with CRT).
In this way, if I configure CMake to look for static linking with boost, it cannot find the libraries.

What is the original reason of this renaming? Since CMake looks for these filenames, while should we break it?

**Comment by ras0219-msft** [comment]
Sorry for taking so long to respond, and thank you for bringing up this issue!

We're definitely aware of the various issues around boost here (#369, #392, #601, #694). The essence of the problem (and solution) is:
>vcpkg only allows one instance of each library to be installed at a time. Therefore, the names of the lib files should not matter; the "capability" of boost is either there or it is not.
>
>Furthermore, a particular triplet can be configured to build some libraries statically and some dynamically. For example, you may opt to create your own triplet that builds zlib as a dll, and then all other dependencies are built as static libs. A portfile of another library can't easily know what configuration the other libraries were built with.
>
>Therefore, it can't know whether boost is static or dynamic in order to pass the appropriate flag into FindBOOST. We tried stripping out the build-specific identifiers from the names, but it did not work out (#369). Therefore, we rename the static names to the dynamic names, and then always look for the dynamic names. Therefore, you always use FindBOOST in the same way and the lib file does the right thing.

So, the current intended behavior is for downstream projects is to _leave all boost settings at their defaults_. Then, you use the Vcpkg triplet to choose what settings you want for boost (static link/static crt/etc) at configure time. On other systems, this can be easily manipulated at configure time via `-DBoost_USE_STATIC_LIBS=ON`.

**Comment by cenit** [comment]
Dear Robert,
thank you very much for your answer. At first I really struggled to understand it: for me, static linking is broken, but for you it is working.
So I read with much more attention your suggestion and found an interesting difference: in my CMakeLists.txt, I am used to use 
```
set(Boost_USE_STATIC_LIBS ON)
```
to request static linking, which is NOT working with vcpkg without my PR (which, to be honest, imho, keeps the downstream project much more consistent with original naming!)

On the other hand, you suggested to use
```
add_definitions(-D_Boost_USE_STATIC_LIBS=ON)
```
which, surprisingly, WORKS with vcpkg to static link Boost libraries.
Is this an intended behaviour?

**Comment by paul-michalik** [comment]
Sorry but this IS an issue. By renaming the boost static libraries the many years lasting convention is broken and all tools on this planet break. So please remove the renaming section.

**Comment by cenit** [comment]
I am waiting on Robert. My pull request was closed, so I don't have big hopes on it being fixed soon

**Comment by muemart** [comment]
AFAIU, the issue is how CMake's FindBoost works.
When libraries want to use boost, most of them just write `find_package(Boost)`, but this by default only finds the dynamic libraries. This is bad for vcpkg, because boost might have been build statically, and then boost doesn't get found. The workaround used then is to rename the boost libraries to always have the dynamic names. The solution for users is to always assume the libraries are named as if they have been built dynamically.
@cenit The reason `add_definitions(-D_Boost_USE_STATIC_LIBS=ON)` works is that `_Boost_USE_STATIC_LIBS` doesn't do anything, so the module uses default settings and searches libraries with the dynamic name and finds them.

**Comment by ras0219-msft** [comment]
I've just pushed 63454f61 which should fix this problem by unsetting `Boost_USE_STATIC_LIBS` when calling `find_package(Boost)` and using the Vcpkg toolchain file. You'll properly get a static version of Boost if you use the `-static` triplets and a dynamic version if you use the non-static triplets.

Whenever you're not using Vcpkg, your projects will (of course) be completely unaffected.

**Comment by jtravs** [comment]
Note that this fix does not seem to solve https://github.com/Microsoft/vcpkg/issues/1483

**Comment by paul-michalik** [comment]
@muemart The standard `findBoost` finds the Boost static libraries on Windows **iff** they stick with the expected naming convention. The same applies to the Boost auto-linking feature. You can't break the naming convention and expect that things keep working. My proposal is just not to rename the products of the bjam build at all. In my fork of [vcpkg](https://github.com/paul-michalik/vcpkg) I have implemented this modification and use it for my custom triplets which are set up to prefer building static libraries if possible. Please let me know what's wrong with this approach! Moreover, if there is a technical or other reason for this renaming, I'd consider it a fundamental flaw in vcpkg' workflow which needs to be fixed.

**Comment by paul-michalik** [comment]
One more note: I'd consider a flawless build and deployment of Boost libraries a super-high priority feature, since this will very likely be the first dependency every developer will test. If that doesn't work out of the box, chances are that vcpkg won't make it into the toolbox at all... which I'd consider a pity.

**Comment by muemart** [comment]
The problem is that the following CMake file will *not* find the libboost_... libraries.
```
cmake_minimum_required(VERSION 3.8)
project(dummy)
find_package(Boost REQUIRED iostreams)
```
But this is how most projects use it. That's bad for vcpkg, because boost might be built statically in the current triplet. Even worse, users can write custom triplets where one library is static and another one dynamic, so build systems can't be patched to look for the static boost libraries, because the ports are independent and a portfile doesn't know if it should look for boost_... or libboost_.... Renaming libraries is a straightforward way to fix this.

**Comment by ras0219-msft** [comment]
@paul-michalik You're absolutely right that Boost is a top priority! We need it to work well with every part of the vcpkg experience (MSBuild, CMake, and other libraries inside vcpkg).

With the change I've made in 63454f6, the following CMake works for both `x86-windows` and `x64-windows-static`:
```cmake
project(dummy)
#The following line can be commented OR uncommented
#set(Boost_USE_STATIC_LIBS ON)
find_package(Boost REQUIRED COMPONENTS filesystem)
```

In each of the four cases, the type of boost will be determined by the triplet (i.e. it will always be static in the static triplet and always dynamic in the dynamic triplet).

**Comment by paul-michalik** [comment]
@ras0219-msft, @muemart: For the dummy project, this is exactly the behavior I'd expect, if the Boost default (= use dynamic libraries) is defined. I'd also expect CMake to throw an error, if only static libraries are present on Windows and Boost_USE_STATIC_LIBS = OFF . This is documented and well known feature.

@ras0219-msft Hm, but this assumes that users are using the vcpkg.cmake toolchain file which isn't working for custom triplets, for example. Also, frankly, I dislike the misuse of the toolchain file for this.

**Comment by cenit** [comment]
not sure that bypassing explicit requests in the CMakeLists.txt is appropriate, are we sure about not causing bigger problems elsewhere?

**Comment by ras0219-msft** [comment]
@paul-michalik If our toolchain doesn't support custom triplets, that's a serious bug we need to fix. However, in my tests, it appears to work fine:
```
cmake .. -DCMAKE_TOOLCHAIN_FILE=.../vcpkg.cmake -DVCPKG_TARGET_TRIPLET=x64-windows-v140
```

The logic behind our integration as a toolchain file is that we believe vcpkg is best viewed as part of the "target system" that you're "cross compiling for". Each triplet matches a different target system type.

@cenit As far as we can tell, moving the deployment choice into vcpkg and out of the build script does not cause any problems. Though, if we do find any, we should fix them.

**Comment by Warix3** [comment]
Thank you for making this rename thing and wasting hours of my time.

**Comment by MikeGitb** [comment]
@ras0219-msft: Sorry for digging up this old thread, but what exactly is the problem that is solved by the renaming? From how I understand the previous posts from you and muemart, the goal seems to be to have a workaround for people who build only the static boost libraries, but then expect `find_package(Boost)` to work without `set(Boost_USE_STATIC_LIBS ON)` is that correct? What is the "usual" cmake behavior here? From paul-michalik's answer I get the impression, that this is actually not supposed to work that way anyway.

If so, I think it is really a bad idea to try to silently "fix" broken cmake scripts for the user, because it invalidates a lot of documentation out there and might potentially even lead to compatibility issues. It's a bit like a compiler that accepts invalid c++ constructs by guessing what the user wants instead of  throwing an error and telling the User that he made a mistake. 

Btw., I came here because I was wondering, why the .lib files apparently got compiled with a different toolchain than the dll files (v140 vs v141)

**Comment by ras0219-msft** [comment]
Here's the summary of the issues:
- Many libraries' cmake scripts define Boost_USE_STATIC_LIBS on Windows and have no way to turn it off. This is to make things work "out of the box" with the precompiled boost distribution, however we believe the choice of using static libraries or dynamic libraries should be left to the final program assembler (i.e. you) instead of the intermediate libraries. This is already specified via the triplet you're using, so we "type-erase" this setting.
- Boost's naming convention considers v140 and v141 binary incompatible, even though they **are** compatible. This isn't really a fault on their part -- this is the first time MSVC has done so in my memory! However, we want to be able to build boost with VS2017 even if you're consuming from VS2015 (or vice versa), so we need to "type-erase" the compiler.
- Finally, CMake < 3.10 doesn't handle the new architecture flags, but CMake > 3.10 doesn't handle _not_ having the architecture flags. The only way to satisfy both is to "type-erase" the version, which makes them both handle things correctly. They pick up dependency information from the `version.hpp`, so this doesn't negatively affect the target dependency information (boost::system, boost::date_time, etc).

We currently type-erase all the above by always generating a single naming scheme for the _import_ libraries (dynamic linking, v140, no version, no arch, dynamic crt). Then, when using `find_package(Boost)` we intercept the call and set all the correct search parameters to find these libs.

Note that the DLLs still have the original names, so your _deployment_ should be unaffected (it can live side-by-side with older boost versions, etc).

An alternative way to achieve this would be to record the settings used during the build, then have a way to pipe those into the `find_package(Boost)` call. This is arguably more "pure", but it is also more complex. Note though that this still wouldn't fix the CMake issue above. The primary advantage of this approach would be compatibility with the _default_ `auto_link.hpp`.

However, we _can_ fix auto_link in the current scheme! We just need to do a relatively straightforward string replace into `auto_link.hpp` which encodes our target naming scheme into the header so that it will find the correct binaries as well. I'd love to accept a PR doing this and it's on the backlog for us to investigate as well.

**Comment by MikeGitb** [comment]
Thanks for the detailed explanation.

**Comment by paul-michalik** [comment]
In the meantime my strategy for dealing with the "Boost issue" is this - ignore it and trust the vcpkg workaround. I switched to using the provided toolchain file and standard triplets for which the solution works out of the box. My only addition to the toolchain is a handcrafted  `vcpkg.props` which I use for MSBuild projects instead of the generated NuGet packages or system wide integration.

**Comment by hellcats** [comment]
I've read through all this stuff and still I'm completely confused.  I'm on Win 10-64bit and have installed the latest vcpkg and cmake.  I'm trying the following simple CMakeLists.txt

```
cmake_minimum_required(VERSION 3.0)
project(test)
find_package(Boost REQUIRED)
add_executable(main main.cpp)
```

I run cmake with:
`cmake -G "Visual Studio 15 2017 Win64" .. -DCMAKE_TOOLCHAIN_FILE=C:/code/vcpkg/scripts/buildsystems/vcpkg.cmake`

And I always get errors:
```
CMake Error at C:/Program Files/CMake/share/cmake-3.10/Modules/FindBoost.cmake:1928 (message):
  Unable to find the requested Boost libraries.

  Unable to find the Boost header files.  Please set BOOST_ROOT to the root
  directory containing Boost or BOOST_INCLUDEDIR to the directory containing
  Boost's headers.
Call Stack (most recent call first):
  C:/code/vcpkg/scripts/buildsystems/vcpkg.cmake:196 (_find_package)
  CMakeLists.txt:16 (find_package)
```
I'm really just about to give up on vcpkg

**Comment by cenit** [comment]
since you are not using default triplet, you also have to pass on the command line `"-DVCPKG_TARGET_TRIPLET=x64-windows"`

**Comment by hellcats** [comment]
@cenit Your hint got me 99% of the way; I only had to add -static to the triplet and it worked! (e.g. `"-DVCPKG_TARGET_TRIPLET=x64-windows-static"`). I was so used to seeing the error message it took me a few seconds to realize it had worked :)  So I guess I need to read more about these configuration triplets...

**Comment by cenit** [comment]
glad you are able to work now!

**Comment by mat007** [comment]
So in the end there is no way I can mix in my cmake built project a custom built version of boost living in a totally separate location with other libraries retrieved by vcpkg?
It feels odd that, given I don't even vcpkg install boost, passing -DCMAKE_TOOLCHAIN_FILE=… to cmake undoes set( Boost_USE_STATIC_LIBS ON ) behind my back (it took me ages to pinpoint this 😄 ).

**Comment by redbaron** [comment]
For those who is pulling their hair trying to statically link vcpkg installed boost when building external (non-vcpkg) projects like me here is recipe, which is never mentioned in this thread:

 1. You'll need CMake >= 3.18
 1. Install boost static lib: `vcpkg.exe install boost:x64-windows-static-md`
 1. in your cmake project use these:
    ```
    cmake_minimum_required(VERSION 3.18.0)  # that's where support for Boost_LIB_PREFIX was added
    set(Boost_LIB_PREFIX "")                # this line cost me 4 hours of my life
    ```
  1. Not required, because FindBoost has tons of workarounds for all sorts of quirks, but IMHO nice to have for clarity:
      ```
      set(Boost_USE_STATIC_LIBS ON)
      set(Boost_USE_STATIC_RUNTIME OFF)   # matches `windows-static-md` where CRT is dynamically linked
      ```
